src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNMethodContext.java

```diff
@@ -0,0 +1,171 @@
+import static com.amazon.opendistroforelasticsearch.knn.common.KNNConstants.NAME;
+import static com.amazon.opendistroforelasticsearch.knn.common.KNNConstants.PARAMETERS;
+
+/**
+ * KNNMethodContext will contain the information necessary to produce a library index from an Opensearch mapping.
+ * It will encompass all parameters necessary to build the index.
+ */
+public class KNNMethodContext implements ToXContentFragment {
+
+    public static final KNNMethodContext DEFAULT = new KNNMethodContext(KNNEngine.DEFAULT, SpaceType.DEFAULT,
+            new MethodComponentContext(METHOD_HNSW, Collections.emptyMap()));
+
+    private KNNEngine knnEngine;
+    private SpaceType spaceType;
+    private MethodComponentContext methodComponent;
+
+    /**
+     * Constructor
+     *
+     * @param knnEngine engine that this method uses
+     * @param spaceType space type that this method uses
+     * @param methodComponent MethodComponent describing the main index
+     */
+    public KNNMethodContext(KNNEngine knnEngine, SpaceType spaceType, MethodComponentContext methodComponent) {
+        this.knnEngine = knnEngine;
+        this.spaceType = spaceType;
+        this.methodComponent = methodComponent;
+    }
+
+    /**
+     * Gets the main method component
+     *
+     * @return methodComponent
+     */
+    public MethodComponentContext getMethodComponent() {
+        return methodComponent;
+    }
+
+    /**
+     * Gets the engine to be used for this context
+     *
+     * @return knnEngine
+     */
+    public KNNEngine getEngine() {
+        return knnEngine;
+    }
+
+    /**
+     * Gets the space type for this context
+     *
+     * @return spaceType
+     */
+    public SpaceType getSpaceType() {
+        return spaceType;
+    }
+
+    /**
+     * This method uses the knnEngine to validate that the method is compatible with the engine
+     *
+     */
+    public void validate() {
+        knnEngine.validateMethod(this);
+    }
+
+    /**
+     * Parses an Object into a KNNMethodContext.
+     *
+     * @param in Object containing mapping to be parsed
+     * @return KNNMethodContext
+     */
+    public static KNNMethodContext parse(Object in) {
+        if (!(in instanceof Map)) {
+            throw new MapperParsingException("Unable to parse mapping into KNNMethodContext. Object not of type \"Map\"");
+        }
+
+        @SuppressWarnings("unchecked")
+        Map<String, Object> methodMap = (Map<String, Object>) in;
+
+        KNNEngine engine = KNNEngine.DEFAULT;
+        SpaceType spaceType = SpaceType.DEFAULT;
+        String name = "";
+        Map<String, Object> parameters = null;
+
+        String key;
+        Object value;
+        for (Map.Entry<String, Object> methodEntry : methodMap.entrySet()) {
+            key = methodEntry.getKey();
+            value = methodEntry.getValue();
+            if (KNN_ENGINE.equals(key)) {
```

Comment:
can you refactor this into method? It is very hard to understand.

Reply:
Which portion into a method? Basically, we are looping over each key, and value in the user provided map and getting and validating the value they pass in. 

I could add additional comments, but Im not sure I see how to improve via a method.

Reply:
IMO, Since it is not an array, we don't have to iterate, instead we can get/validate/transform. I posted suggestion on what i meant, feel free to update if logic is not valid.

Reply:
Right, I thought about this. My only concern is that without iterating, we allow a user to pass in parameters that do not do anything. Hence, in the else check, we are able to throw error if an invalid key is passed in.